libraries/rushell/src/AstNode.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Token } from './Tokenizer';
+
+export enum AstKind {
+  Script,
+  AndIf,
+  Command,
+  CompoundWord,
+  VariableExpansion,
+  Text
+}
+
+/**
+ * Base class for all AST nodes.
+ */
+export abstract class AstNode {
+  public abstract get kind(): AstKind;
+
+  /**
+   * Returns a diagnostic dump of the tree, showing the prefix/suffix/separator for
+   * each node.
+   */
+  public getDump(indent: string = ''): string {
+    let result: string = indent + AstKind[this.kind];
+
+    const dumpText: string | undefined = this.getDumpText();
+    if (dumpText) {
+      result += '=' + JSON.stringify(dumpText);
+    }
+
+    const childNodes: AstNode[] = this.getChildNodes();
+    if (childNodes.length === 0) {
+      result += '\n';
+    } else {
+      result += ':\n';
+      for (const child of this.getChildNodes()) {
+        result += child.getDump(indent + '  ');
+      }
+    }
+
+    return result;
+  }
+
+  public getChildNodes(): AstNode[] {
+    const nodes: AstNode[] = [];
+    this.pushChildNodes(nodes);
+    return nodes;
+  }
+
+  protected abstract pushChildNodes(nodes: AstNode[]): void;
```

Comment:
>pushChildNodes [](start = 21, length = 14)

The name of this function makes me expect that you call it to add child nodes, not to get them back out.